light-client/eth-light-client/src/server.rs

```diff
@@ -0,0 +1,92 @@
+use ethers::{abi::AbiEncode, types::H256};
+use eyre::{Report, Result};
+use salvo::prelude::*;
+use serde::Deserialize;
+use serde_json::json;
+
+use crate::{config::Config, db::DB, merkle};
+
+#[derive(Deserialize)]
+struct RootReq {
+	block_number: u64,
+}
+
+#[derive(Deserialize)]
+struct VerifyReq {
+	block_number: u64,
+	hashes: Vec<H256>,
+	indices: Vec<u32>,
+}
+
+#[handler]
+async fn status() -> &'static str {
+	"ok"
+}
+
+#[handler]
+async fn root(req: &mut Request, dep: &mut Depot) -> eyre::Result<String> {
+	let root_req = req
+		.parse_body::<RootReq>()
+		.await
+		.map_err(|_| Report::msg("Could not parse RootReq"))?;
+	let db = dep.obtain::<DB>().ok_or(Report::msg("Could not get DB"))?;
+	let logs = db
+		.select_logs(root_req.block_number)
+		.map_err(|_| Report::msg("Could not get logs"))?;
+	let hashes = logs
+		.iter()
+		.flat_map(|log| log.transaction_hash)
+		.collect::<Vec<_>>();
+
+	Ok(json!({
+		"root": merkle::root(&hashes).encode_hex()
+	})
+	.to_string())
+}
```

Comment:
So the Merkle root which contains the logs is actually part of the block, and is not computed this way. They are part of the "Receipts tree".

![receipts tree](https://raw.githubusercontent.com/zmitton/eth-proof/master/img/architecture-diagram.JPG)

Here is a javascript implementation for getting receipts... https://github.com/syscoin/eth-proof

Reply:
Now, I use transaction receipts to get transaction logs.